packages/wasm/src/runtime/instance.rs

```diff
@@ -64,29 +58,28 @@ impl State {
 pub struct WasmInstance {
     instance: Instance,
     pub store: Store,
-    pub module: Module,
 }
 
 impl WasmInstance {
-    pub fn new(wasm_module: &[u8], state: Arc<Mutex<State>>) -> Result<Self, WrapperError> {
+    pub fn new(
+        module: &Module,
+        memory_initial_limits: u8,
+        state: Arc<Mutex<State>>,
+    ) -> Result<Self, WrapperError> {
         let mut store = Store::default();
-        let module = Module::new(&store, wasm_module).unwrap();
-        let memory = WasmInstance::create_memory(&mut store, wasm_module)?;
-        let imports = create_imports(memory.clone(), &mut store, state.clone());
+        let memory = WasmInstance::create_memory(&mut store, memory_initial_limits)?;
+
+        state.lock().unwrap().memory = Some(memory.clone());
```

Comment:
I think we want to get rid of this clone, no? I would think it makes more sense to do this:
```rust
state.lock().unwrap().memory = WasmInstance::create_memory(
    &mut store,
    memory_initial_limits
)?;

let memory = &state.lock().unwrap().memory;
```

This way we don't copy, and we have a single instance within state & the function's local stack.

Reply:
On this line `let memory = &state.lock().unwrap().memory;` you are moving out of the mutex guard.
I don't think this can be done without copy, or at least I don't see how.

That said, memory appears to be just a handle to real memory, so copying should be fast anyway.

Reply:
Okay perfect, thank you for explaining this :)